src/Hazelcast.Net/CP/CpSubsystemSession.cs

```diff
@@ -0,0 +1,271 @@
+        private readonly ConcurrentDictionary<CPGroupId, CPSubsystemSessionState> _sessions = new ConcurrentDictionary<CPGroupId, CPSubsystemSessionState>();
+        //Note: threadId is bounded to context of task(AsyncLocal) not to thread(ThreadLocal) that runs it.
+        private readonly ConcurrentDictionary<(CPGroupId, long), long> _threadIds = new ConcurrentDictionary<(CPGroupId, long), long>();
+        private int _disposed;
+        public const int NoSessionId = -1;
+        private object _mutex = new object();
+        private SemaphoreSlim _semaphoreReadWrite = new SemaphoreSlim(1, 1);
+
+        private readonly ILogger _logger;
+
+        //Cluster Fields
+        private readonly Cluster Cluster;
+        private readonly ClusterState State;
+        #endregion
+
+        #region SessionManagement
+        public CPSubsystemSession(Cluster cluster)
+        {
+            Cluster = cluster ?? throw new ArgumentNullException(nameof(cluster));
+            State = cluster.State ?? throw new ArgumentNullException(nameof(cluster.State));
+
+            _logger = State.LoggerFactory.CreateLogger<CPSubsystemSession>();
+            _cancel = new CancellationTokenSource();
+            HConsole.Configure(x => x.Configure<Heartbeat>().SetPrefix("CP.HEARTBEAT"));
+        }
+
+        /// <summary>
+        /// Acquires the session by increasing given count, creates if absent.
+        /// </summary>
+        /// <param name="groupId"></param>
+        /// <param name="count">Increase count of acquirment</param>
+        /// <returns>Session Id</returns>
+        public async Task<long> AcquireSessionAsync(CPGroupId groupId, int count = 1)
+        {
+            var session = await GetOrCreateSessionAsync(groupId).CfAwait();
+            session.Acquire(count);
+            return session.Id;
+        }
+
+        /// <summary>
+        /// Releases the session by decreasing given count
+        /// </summary>
+        /// <param name="groupId"></param>
+        /// <param name="sessionId"></param>
+        /// <param name="count">Decrease count of release</param>
+        /// <returns></returns>
+        public void ReleaseSession(CPGroupId groupId, long sessionId, int count = 1)
+        {
+            if (_sessions.TryGetValue(groupId, out var sessionState) && sessionState.Id == sessionId)
+            {
+                sessionState.Release(count);
+            }
+        }
+
+        /// <summary>
+        /// Invalidates the given session after invalidation no more heartbeat will be sent.
+        /// </summary>
+        /// <param name="groupId"></param>
+        /// <param name="sessionId"></param>
+        public void InvalidateSession(CPGroupId groupId, long sessionId)
+        {
+            if (_sessions.TryGetValue(groupId, out var sessionState) && sessionState.Id == sessionId)
+            {
+                _sessions.TryRemove(groupId, out var _);
```

Comment:
java also tries to match the value of the map key to remove, is it possible to do the same?

Reply:
Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/590/commits/615acd74e54cd18425eb94dfeb577ca5512e2a61